src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -1212,8 +1224,15 @@ DatasetClause SparqlQleverVisitor::visit(Parser::UsingClauseContext* ctx) {
 
 // ____________________________________________________________________________________
 void Visitor::visit(Parser::PrologueContext* ctx) {
-  visitVector(ctx->baseDecl());
-  visitVector(ctx->prefixDecl());
+  for (auto* child : ctx->children) {
+    if (auto* baseDecl = dynamic_cast<Parser::BaseDeclContext*>(child)) {
+      visit(baseDecl);
+    } else {
+      auto* prefixDecl = dynamic_cast<Parser::PrefixDeclContext*>(child);
+      AD_CORRECTNESS_CHECK(prefixDecl);
+      visit(prefixDecl);
+    }
+  }
```

Comment:
What is wrong with the simpler older version?

Reply:
Imagine the following case:
```sparql
BASE <http://example.com>
PREFIX test: <test>
BASE <http://other.example.com>
PREFIX test2: <test>
BASE <http://alternative.example.com>
SELECT * FROM {
  <s> ?p test:abc
}
```

Then the first `BASE` should apply to the first prefix, and the second `BASE` should apply to the second prefix, and the third `BASE` should apply to everything else that follows. So to ensure that we apply `BASE` to `PREFIX` correctly, we can't parse them after each other, and instead have to interleave parsing.